modules/ding_seo/ding_seo.module

```diff
@@ -0,0 +1,228 @@
+  // the ting entity, so might as well safe some time by not going through
+  // entity API.
+  /** @var \Ting\TingObjectInterface */
+  $ting_object = ting_search_object_load($ting_object_id);
+  if (isset($ting_object)) {
+    /** @var DingSEO\TingObjectSchemaWrapperInterface */
+    $schema_wrapper = ding_seo_get_ting_object_schema_wrapper($ting_object);
+  }
+
+  // Bail out if we were not able to produce a schema wrapper or the schema
+  // type could not be determined.
+  $schema_type = NULL;
+  if (!isset($schema_wrapper) || !$schema_type = $schema_wrapper->getSchemaType()) {
+    print '';
+    drupal_exit();
+  }
+
+  // Common properties.
+  $node = [
+    '@type' => $schema_type,
+    'name' => $schema_wrapper->getName(),
+    'description' => $schema_wrapper->getDescription(),
+  ];
+  if ($image_url = $schema_wrapper->getImageURL()) {
+    $node['image'] = [
+      '@type' => 'ImageObject',
+      'representativeOfPage' => 'True',
+      'url' => $image_url,
+    ];
+
+    // TODO: Find out if it's ok to omit dimensions or we should just ditch
+    // it all together if not found. We know there's a cover, but it might
+    // not be fecthed yet so we can check dimensions.
+    if ($image_dimensions = $schema_wrapper->getImageDimensions()) {
+      $node['image'] += [
+        'width' => "$image_dimensions[0]",
+        'height' => "$image_dimensions[1]",
+      ];
+    }
+  }
+
+  switch ($schema_type) {
+    case 'Book':
+      $authors = $schema_wrapper->getAuthors();
+      $node += [
+        '@id' => $schema_wrapper->getCollectionURL(),
+        'url' => $schema_wrapper->getCollectionURL(),
+        'author' => ding_seo_get_person_nodes($authors),
+      ];
+
+      foreach ($schema_wrapper->getWorkExamples() as $work_example_wrapper) {
+        $work_example = [
+          '@type' => 'Book',
+          '@id' => $work_example_wrapper->getObjectURL(),
+          'url' => $work_example_wrapper->getObjectURL(),
+          'name' => $work_example_wrapper->getName(),
+          'bookEdition' => $work_example_wrapper->getBookEdition(),
+          'datePublished' => $work_example_wrapper->getDatePublished(),
+          'isbn' => $work_example_wrapper->getISBN(),
+          'bookFormat' => $work_example_wrapper->getBookFormat(),
+        ];
+
+        $work_example_authors = $work_example_wrapper->getAuthors();
+        // Only include author information if it's different than the
+        // information on the top level Book entity.
+        if ($work_example_authors != $authors) {
+          $work_example['author'] = ding_seo_get_person_nodes($work_example_authors);
+        }
+        // Filter out any empty fields before adding to node.
+        $node['workExample'][] = array_filter($work_example);
+      }
+      break;
+    case 'Movie':
+      $node += [
+        '@id' => $schema_wrapper->getObjectURL(),
+        'url' => $schema_wrapper->getObjectURL(),
```

Comment:
I have moved more of the properties to common.

We still need some special handling for books since they have nested work examples with some additional props and some them should be repeated if they differ from top-level work.

Also different kind of contributors for movies and books.